gradle action fails if gradlew is in subdirectory

In this project, `gradlew` is in `$SOURCE_DIR/android/gradlew` and the fastlane configuration files in `$SOURCE_DIR/fastlane`. The `gradle` action is called like this:

~~~~
gradle(
        gradle_path: "android/gradlew",
        task: "lint#{app_variant}Release",
        flags: "--info"
      )
~~~~

Expected: `gradlew` should be called with `android` as its working directory.

~~~~
/Users/sk/.rvm/gems/ruby-2.1.1/gems/fastlane-1.34.0/lib/fastlane/helper/sh_helper.rb:36:in `sh_no_action': Exit status of command 'android/gradlew lintLocosonicPlayerRelease --info' was 1 instead of 0.  (RuntimeError)
Starting Build
Settings evaluated using empty settings script.
Projects loaded. Root project using empty build file.
Included projects: [root project 'locosonic-player']
Evaluating root project 'locosonic-player' using empty build file.
All projects evaluated.

FAILURE: Build failed with an exception.

* What went wrong:
Task 'lintLocosonicPlayerRelease' not found in root project 'locosonic-player'.

* Try:
Run gradlew tasks to get a list of available tasks. Run with --stacktrace option to get the stack trace. Run with --debug option to get more log output.

BUILD FAILED

Total time: 1.159 secs
	from /Users/sk/.rvm/gems/ruby-2.1.1/gems/fastlane-1.34.0/lib/fastlane/helper/sh_helper.rb:8:in `sh'
	from /Users/sk/.rvm/gems/ruby-2.1.1/gems/fastlane-1.34.0/lib/fastlane/action.rb:67:in `sh'
	from /Users/sk/.rvm/gems/ruby-2.1.1/gems/fastlane-1.34.0/lib/fastlane/helper/gradle_helper.rb:30:in `trigger'
	from /Users/sk/.rvm/gems/ruby-2.1.1/gems/fastlane-1.34.0/lib/fastlane/actions/gradle.rb:14:in `run'
	from /Users/sk/.rvm/gems/ruby-2.1.1/gems/fastlane-1.34.0/lib/fastlane/runner.rb:142:in `block (2 levels) in execute_action'
	from /Users/sk/.rvm/gems/ruby-2.1.1/gems/fastlane-1.34.0/lib/fastlane/actions/actions_helper.rb:37:in `execute_action'
	from /Users/sk/.rvm/gems/ruby-2.1.1/gems/fastlane-1.34.0/lib/fastlane/runner.rb:128:in `block in execute_action'
	from /Users/sk/.rvm/gems/ruby-2.1.1/gems/fastlane-1.34.0/lib/fastlane/runner.rb:127:in `chdir'
	from /Users/sk/.rvm/gems/ruby-2.1.1/gems/fastlane-1.34.0/lib/fastlane/runner.rb:127:in `execute_action'
	from /Users/sk/.rvm/gems/ruby-2.1.1/gems/fastlane-1.34.0/lib/fastlane/fast_file.rb:149:in `method_missing'
	from Fastfile:249:in `block (3 levels) in parsing_binding'
	from Fastfile:58:in `call'
	from Fastfile:58:in `mk_production_lane'
	from Fastfile:248:in `block (2 levels) in parsing_binding'
	from /Users/sk/.rvm/gems/ruby-2.1.1/gems/fastlane-1.34.0/lib/fastlane/lane.rb:36:in `call'
	from /Users/sk/.rvm/gems/ruby-2.1.1/gems/fastlane-1.34.0/lib/fastlane/lane.rb:36:in `call'
	from /Users/sk/.rvm/gems/ruby-2.1.1/gems/fastlane-1.34.0/lib/fastlane/runner.rb:51:in `block in execute'
	from /Users/sk/.rvm/gems/ruby-2.1.1/gems/fastlane-1.34.0/lib/fastlane/runner.rb:46:in `chdir'
	from /Users/sk/.rvm/gems/ruby-2.1.1/gems/fastlane-1.34.0/lib/fastlane/runner.rb:46:in `execute'
	from /Users/sk/.rvm/gems/ruby-2.1.1/gems/fastlane-1.34.0/lib/fastlane/lane_manager.rb:43:in `cruise_lane'
	from /Users/sk/.rvm/gems/ruby-2.1.1/gems/fastlane-1.34.0/lib/fastlane/command_line_handler.rb:31:in `handle'
	from /Users/sk/.rvm/gems/ruby-2.1.1/gems/fastlane-1.34.0/bin/fastlane:35:in `block (2 levels) in run'
	from /Users/sk/.rvm/gems/ruby-2.1.1/gems/commander-4.3.5/lib/commander/command.rb:178:in `call'
	from /Users/sk/.rvm/gems/ruby-2.1.1/gems/commander-4.3.5/lib/commander/command.rb:178:in `call'
	from /Users/sk/.rvm/gems/ruby-2.1.1/gems/commander-4.3.5/lib/commander/command.rb:153:in `run'
	from /Users/sk/.rvm/gems/ruby-2.1.1/gems/commander-4.3.5/lib/commander/runner.rb:428:in `run_active_command'
	from /Users/sk/.rvm/gems/ruby-2.1.1/gems/commander-4.3.5/lib/commander/runner.rb:68:in `run!'
	from /Users/sk/.rvm/gems/ruby-2.1.1/gems/commander-4.3.5/lib/commander/delegates.rb:15:in `run!'
	from /Users/sk/.rvm/gems/ruby-2.1.1/gems/fastlane-1.34.0/bin/fastlane:152:in `run'
	from /Users/sk/.rvm/gems/ruby-2.1.1/gems/fastlane-1.34.0/bin/fastlane:158:in `<top (required)>'
	from /Users/sk/.rvm/gems/ruby-2.1.1/bin/fastlane:23:in `load'
	from /Users/sk/.rvm/gems/ruby-2.1.1/bin/fastlane:23:in `<main>'
	from /Users/sk/.rvm/gems/ruby-2.1.1/bin/ruby_executable_hooks:15:in `eval'
	from /Users/sk/.rvm/gems/ruby-2.1.1/bin/ruby_executable_hooks:15:in `<main>'
~~~~

--- a/docs/Actions.md
+++ b/docs/Actions.md
@@ -89,6 +89,44 @@ Use `gym --help` to get all available options.
 
 The alternative to `gym` is [`ipa`](#ipa) which uses [shenzhen](https://github.com/nomad/shenzhen) under the hood.
 
+### gradle
+
+You can use the `gradle` action to integrate your gradle tasks into `fastlane`.
+
+Simple usage:
+```ruby
+gradle(
+  task: 'assemble',
+  flavor: 'WorldDomination',
+  build_type: 'Release'
+)
+```
+
+In case of an `assemble` task, the signed apk path is accessible in: `Actions.lane_context[Actions::SharedValues::GRADLE_APK_OUTPUT_PATH]`
+
+
+You can pass [gradle properties](https://docs.gradle.org/current/userguide/build_environment.html):
+```ruby
+gradle(
+  # ...
+
+  properties: {
+    'versionCode' => 100,
+    'versionName' => '1.0.0',
+    # ...
+  }
+)
+```
+
+To pass any other CLI flags to gradle use:
+```ruby
+gradle(
+  # ...
+
+  flags: "--exitcode --xml file.xml"
+)
+```
+
 ### verify_xcode
 
 Verifies that the Xcode installation is properly signed by Apple. This is relevant after recent [attacks targeting Xcode](http://researchcenter.paloaltonetworks.com/2015/09/novel-malware-xcodeghost-modifies-xcode-infects-apple-ios-apps-and-hits-app-store/).

--- a/lib/fastlane/actions/gradle.rb
+++ b/lib/fastlane/actions/gradle.rb
@@ -1,36 +1,77 @@
+require 'pathname'
+require 'shellwords'
+
 module Fastlane
   module Actions
     module SharedValues
-      GRADLE_APK_OUTPUT_PATH = :APK_OUTPUT_PATH
+      GRADLE_APK_OUTPUT_PATH = :GRADLE_APK_OUTPUT_PATH
+      GRADLE_ALL_APK_OUTPUT_PATHS = :GRADLE_ALL_APK_OUTPUT_PATHS
       GRADLE_FLAVOR = :GRADLE_FLAVOR
+      GRADLE_BUILD_TYPE = :GRADLE_BUILD_TYPE
     end
 
     class GradleAction < Action
       def self.run(params)
         task = params[:task]
+        flavor = params[:flavor]
+        build_type = params[:build_type]
 
-        gradle = Helper::GradleHelper.new(gradle_path: params[:gradle_path])
+        gradle_task = [task, flavor, build_type].join
 
-        result = gradle.trigger(task: task, flags: params[:flags], serial: params[:serial])
+        project_dir = params[:project_dir]
 
-        return result unless task.start_with?("assemble")
+        gradle_path_param = params[:gradle_path] || './gradlew'
 
-        # We built our app. Store the path to the apk
-        flavor = task.match(/assemble(\w*)/)
-        if flavor and flavor[1]
-          flavor = flavor[1].downcase # Release => release
-          apk_path = Dir[File.join("app", "build", "outputs", "apk", "*-#{flavor}.apk")].last
-          if apk_path
-            Actions.lane_context[SharedValues::GRADLE_APK_OUTPUT_PATH] = File.expand_path(apk_path)
-          else
-            Helper.log.info "Couldn't find signed apk file at path '#{apk_path}'...".red
-            if flavor == 'release'
-              Helper.log.info "Make sure to enable code signing in your gradle task: ".red
-              Helper.log.info "https://stackoverflow.com/questions/18328730/how-to-create-a-release-signed-apk-file-using-gradle".red
-            end
-          end
-          Actions.lane_context[SharedValues::GRADLE_FLAVOR] = flavor
+        # Get the path to gradle, if it's an absolute path we take it as is, if it's relative we assume it's relative to the project_dir
+        gradle_path = if Pathname.new(gradle_path_param).absolute?
+                        File.expand_path(gradle_path_param)
+                      else
+                        File.expand_path(File.join(project_dir, gradle_path_param))
+                      end
+
+        # Ensure we ended up with a valid path to gradle
+        raise "Couldn't find gradlew at path '#{File.expand_path(gradle_path)}'".red unless File.exist?(gradle_path)
+
+        # Construct our flags
+        flags = []
+        flags << "-p #{project_dir.shellescape}"
+        flags << params[:properties].map { |k, v| "-P#{k}=#{v}" }.join(' ') unless params[:properties].nil?
+        flags << params[:flags] unless params[:flags].nil?
+
+        # Run the actual gradle task
+        gradle = Helper::GradleHelper.new(gradle_path: gradle_path)
+
+        # If these were set as properties, then we expose them back out as they might be useful to others
+        Actions.lane_context[SharedValues::GRADLE_BUILD_TYPE] = build_type if build_type
+        Actions.lane_context[SharedValues::GRADLE_FLAVOR] = flavor if flavor
+
+        # Are we assembling, and therefore expecting apk's?
+        is_assembling = task.start_with?('assemble')
+
+        # We need to first clean up any old apk's that might still be lying around because otherwise there is no way of knowing which were the newly generated apk's
+        if is_assembling
+          apk_search_path = File.join(project_dir, '*', 'build', 'outputs', 'apk', '*-*.apk')
+          Dir[apk_search_path].each(&File.method(:delete))
         end
+
+        # We run the actual gradle task
+        result = gradle.trigger(task: gradle_task, serial: params[:serial], flags: flags.join(' '))
+
+        # If we didn't build, then we return now, as it makes no sense to search for apk's in a non-`assemble` scenario
+        return result unless is_assembling
+
+        # Our apk is now built, but there might actually be multiple ones that were built if a flavor was not specified in a multi-flavor project (e.g. `assembleRelease`), however we're not interested in unaligned apk's...
+        new_apks = Dir[apk_search_path].reject { |path| path =~ /^.*-unaligned.apk$/i}
+
+        # We expose all of these new apk's
+        Actions.lane_context[SharedValues::GRADLE_ALL_APK_OUTPUT_PATHS] = new_apks
+
+        # We also take the most recent apk to return as SharedValues::GRADLE_APK_OUTPUT_PATH, this is the one that will be relevant for most projects that just build a single build variant (flavor + build type combo). In multi build variants this value is undefined
+        last_apk_path = new_apks.sort_by(&File.method(:mtime)).last
+        Actions.lane_context[SharedValues::GRADLE_APK_OUTPUT_PATH] = File.expand_path(last_apk_path) if last_apk_path
+
+        # Give a helpful message in case there were no new apk's. Remember we're only running this code when assembling, in which case we certainly expect there to be an apk
+        Helper.log.info 'Couldn\'t find any new signed apk files...'.red if new_apks.empty?
 
         return result
       end
@@ -40,54 +81,75 @@ module Fastlane
       #####################################################
 
       def self.description
-        "All gradle related actions, including building and testing your Android app"
+        'All gradle related actions, including building and testing your Android app'
       end
 
       def self.details
         [
-          "Run `./gradlew tasks` to get a list of all available gradle tasks for your project"
+          'Run `./gradlew tasks` to get a list of all available gradle tasks for your project'
         ].join("\n")
       end
 
       def self.available_options
         [
-          FastlaneCore::ConfigItem.new(key: :serial,
-                                       env_name: "FL_ANDROID_SERIAL",
-                                       description: "Android serial, wich device should be used for this command",
-                                       is_string: true,
-                                       default_value: ""),
           FastlaneCore::ConfigItem.new(key: :task,
-                                       env_name: "FL_GRADLE_TASK",
-                                       description: "The gradle task you want to execute",
+                                       env_name: 'FL_GRADLE_TASK',
+                                       description: 'The gradle task you want to execute, e.g. `assemble` or `test`. For tasks such as `assembleMyFlavorRelease` you should use gradle(task: \'assemble\', flavor: \'Myflavor\', build_type: \'Release\')',
+                                       optional: false,
                                        is_string: true),
-          FastlaneCore::ConfigItem.new(key: :flags,
-                                       env_name: "FL_GRADLE_FLAGS",
-                                       description: "All parameter flags you want to pass to the gradle command, e.g. `--exitcode --xml file.xml`",
+          FastlaneCore::ConfigItem.new(key: :flavor,
+                                       env_name: 'FL_GRADLE_FLAVOR',
+                                       description: 'The flavor that you want the task for, e.g. `MyFlavor`. If you are running the `assemble` task in a multi-flavor project, and you rely on Actions.lane_context[Actions.SharedValues::GRADLE_APK_OUTPUT_PATH] then you must specify a flavor here or else this value will be undefined',
                                        optional: true,
                                        is_string: true),
+          FastlaneCore::ConfigItem.new(key: :build_type,
+                                       env_name: 'FL_GRADLE_BUILD_TYPE',
+                                       description: 'The build type that you want the task for, e.g. `Release`. Useful for some tasks such as `assemble`',
+                                       optional: true,
+                                       is_string: true),
+          FastlaneCore::ConfigItem.new(key: :flags,
+                                       env_name: 'FL_GRADLE_FLAGS',
+                                       description: 'All parameter flags you want to pass to the gradle command, e.g. `--exitcode --xml file.xml`',
+                                       optional: true,
+                                       is_string: true),
+          FastlaneCore::ConfigItem.new(key: :project_dir,
+                                       env_name: 'FL_GRADLE_PROJECT_DIR',
+                                       description: 'The root directory of the gradle project. Defaults to `.`',
+                                       default_value: '.',
+                                       is_string: true),
           FastlaneCore::ConfigItem.new(key: :gradle_path,
-                                       env_name: "FL_GRADLE_PATH",
-                                       description: "The path to your `gradlew`",
+                                       env_name: 'FL_GRADLE_PATH',
+                                       description: 'The path to your `gradlew`. If you specify a relative path, it is assumed to be relative to the `project_dir`',
+                                       optional: true,
+                                       is_string: true),
+          FastlaneCore::ConfigItem.new(key: :properties,
+                                       env_name: 'FL_GRADLE_PROPERTIES',
+                                       description: 'Gradle properties to be exposed to the gradle script',
+                                       optional: true,
+                                       is_string: false),
+          FastlaneCore::ConfigItem.new(key: :serial,
+                                       env_name: 'FL_ANDROID_SERIAL',
+                                       description: 'Android serial, wich device should be used for this command',
                                        is_string: true,
-                                       default_value: Dir["./gradlew"].last, # Using Dir to be nil when the file doesn't exist (import for validation)
-                                       verify_block: proc do |value|
-                                         raise "Couldn't find gradlew at path '#{File.expand_path(value)}'".red unless File.exist?(value)
-                                       end)
+                                       default_value: '')
         ]
       end
 
       def self.output
         [
-          ['GRADLE_APK_OUTPUT_PATH', 'The path to the newly generated apk file']
+          ['GRADLE_APK_OUTPUT_PATH', 'The path to the newly generated apk file. Undefined in a multi-variant assemble scenario'],
+          ['GRADLE_ALL_APK_OUTPUT_PATHS', 'When running a multi-variant `assemble`, the array of signed apk\'s that were generated'],
+          ['GRADLE_FLAVOR', 'The flavor, e.g. `MyFlavor`'],
+          ['GRADLE_BUILD_TYPE', 'The build type, e.g. `Release`']
         ]
       end
 
       def self.return_value
-        "The output of running the gradle task"
+        'The output of running the gradle task'
       end
 
       def self.authors
-        ["KrauseFx"]
+        ['KrauseFx', 'lmirosevic']
       end
 
       def self.is_supported?(platform)

--- a/lib/fastlane/helper/gradle_helper.rb
+++ b/lib/fastlane/helper/gradle_helper.rb
@@ -23,10 +23,11 @@ module Fastlane
       end
 
       # Run a certain action
-      def trigger(task: nil, flags: nil, serial:nil)
+      def trigger(task: nil, flags: nil, serial: nil)
         # raise "Could not find gradle task '#{task}' in the list of available tasks".red unless task_available?(task)
-        android_serial = serial != "" ? "ANDROID_SERIAL=#{serial}" : nil
-        command = [android_serial, gradle_path, task, flags].join(" ")
+
+        android_serial = (serial != "") ? "ANDROID_SERIAL=#{serial}" : nil
+        command = [android_serial, gradle_path, task, flags].reject(&:nil?).join(" ")
         Action.sh(command)
       end
 

--- a/spec/actions_specs/gradle_spec.rb
+++ b/spec/actions_specs/gradle_spec.rb
@@ -2,11 +2,51 @@ describe Fastlane do
   describe Fastlane::FastFile do
     describe "gradle" do
       it "generates a valid command" do
-        result = Fastlane::FastFile.new.parse("lane :test do
-          gradle(task: 'test', gradle_path: './fastlane/README.md')
-        end").runner.execute(:test)
+        result = Fastlane::FastFile.new.parse("lane :build do
+          gradle(task: 'assemble', flavor: 'WorldDomination', build_type: 'Release', properties: { 'versionCode' => 200}, gradle_path: './fastlane/README.md')
+        end").runner.execute(:build)
 
-        expect(result).to eq(" ./fastlane/README.md test ")
+        expect(result).to eq("#{File.expand_path('README.md')} assembleWorldDominationRelease -p . -PversionCode=200")
+      end
+
+      it "correctly uses the serial" do
+        result = Fastlane::FastFile.new.parse("lane :build do
+          gradle(task: 'assemble', flavor: 'WorldDomination', build_type: 'Release', properties: { 'versionCode' => 200}, serial: 'abc123', gradle_path: './fastlane/README.md')
+        end").runner.execute(:build)
+
+        expect(result).to eq("ANDROID_SERIAL=abc123 #{File.expand_path('README.md')} assembleWorldDominationRelease -p . -PversionCode=200")
+      end
+
+      it "supports multiple flavors" do
+        result = Fastlane::FastFile.new.parse("lane :build do
+          gradle(task: 'assemble', build_type: 'Release', gradle_path: './fastlane/README.md')
+        end").runner.execute(:build)
+
+        expect(result).to eq("#{File.expand_path('README.md')} assembleRelease -p .")
+      end
+
+      it "supports multiple build types" do
+        result = Fastlane::FastFile.new.parse("lane :build do
+          gradle(task: 'assemble', flavor: 'WorldDomination', gradle_path: './fastlane/README.md')
+        end").runner.execute(:build)
+
+        expect(result).to eq("#{File.expand_path('README.md')} assembleWorldDomination -p .")
+      end
+
+      it "supports multiple flavors and build types" do
+        result = Fastlane::FastFile.new.parse("lane :build do
+          gradle(task: 'assemble', gradle_path: './fastlane/README.md')
+        end").runner.execute(:build)
+
+        expect(result).to eq("#{File.expand_path('README.md')} assemble -p .")
+      end
+
+      it "supports the backwards compatible syntax" do
+        result = Fastlane::FastFile.new.parse("lane :build do
+          gradle(task: 'assembleWorldDominationRelease', gradle_path: './fastlane/README.md')
+        end").runner.execute(:build)
+
+        expect(result).to eq("#{File.expand_path('README.md')} assembleWorldDominationRelease -p .")
       end
     end
   end

